kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala

```diff
@@ -247,7 +250,7 @@ class KyuubiSessionImpl(
     try {
       if (_client != null) _client.closeSession()
     } finally {
-      if (engine != null) engine.close()
+      openSessionError.foreach { _ => if (engine != null) engine.close() }
```

Comment:
The egine will not be reused if the sharelevel is a CONNECTION and the session is closed, so should close engine whether open session fails or not?

Reply:
With connection share level, engine should close gracefully by itself. Engine exit code will not be 0 if we close engine by Kyuubi Server.

Reply:
Can we add a timeout to wait for the engine status to complete, and then kill the application.

Reply:
We can, but it's a bit complex. I think this pr is enough to fix regression.

I see your requreiements that to avoid the leaked engine. We can improve that in a global view rather than wait and kill one by one. e.g. Use a thread to detect if there is a leaked engine.

Note, it's the key to define what is the `leaked engine`. In general, A `leaked engine` can be:
- status is finished in resource manager but engine build process is alive (for all share level)
- session is closed but status is not fiinshed or engine build process is alive (for connect share level and batch mode)

We can have a new pr if you are interested in.

Reply:
> We can, but it's a bit complex. I think this pr is enough to fix regression.
> 
> I see your requreiements that to avoid the leaked engine. We can improve that in a global view rather than wait and kill one by one. e.g. Use a thread to detect if there is a leaked engine.
> 
> Note, it's the key to define what is the `leaked engine`. In general, A `leaked engine` can be:
> 
> * status is finished in resource manager but engine build process is alive (for all share level)
> * session is closed but status is not fiinshed or engine build process is alive (for connection share level and batch mode)
> 
> We can have a new pr if you are interested in.

I agree with you, this pr is enough to fix the previous issue.